contracts/tatum/Tatum721Provenance.sol

```diff
@@ -0,0 +1,223 @@
+        // saving cashback addresses and values
+        _cashbackRecipients[tokenId] = recipientAddresses;
+        _cashbackValues[tokenId] = cashbackValues;
+        return true;
+    }
+
+    function mintWithTokenURI(
+        address to,
+        uint256 tokenId,
+        string memory uri
+    ) public returns (bool) {
+        require(
+            hasRole(MINTER_ROLE, _msgSender()),
+            "ERC721PresetMinterPauserAutoId: must have minter role to mint"
+        );
+        _mint(to, tokenId);
+        _setTokenURI(tokenId, uri);
+        return true;
+    }
+
+    function supportsInterface(bytes4 interfaceId)
+        public
+        view
+        virtual
+        override(AccessControlEnumerable, ERC721, ERC721Enumerable)
+        returns (bool)
+    {
+        return super.supportsInterface(interfaceId);
+    }
+
+    function tokenURI(uint256 tokenId)
+        public
+        view
+        virtual
+        override(ERC721, ERC721URIStorage)
+        returns (string memory)
+    {
+        return ERC721URIStorage.tokenURI(tokenId);
+    }
+
+    function gettokenData(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (TokenData[] memory)
+    {
+        return _tokenData[tokenId];
+    }
+
+    function caller() public view virtual returns (address) {
+        return _msgSender();
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal virtual override(ERC721, ERC721Enumerable) {
+        super._beforeTokenTransfer(from, to, tokenId);
+    }
+
+    function _burn(uint256 tokenId)
+        internal
+        virtual
+        override(ERC721, ERC721URIStorage)
+    {
+        return ERC721URIStorage._burn(tokenId);
+    }
+
+    function tokenCashbackValues(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (uint256[] memory)
+    {
+        return _cashbackValues[tokenId];
+    }
+
+    function tokenCashbackRecipients(uint256 tokenId)
+        public
+        view
+        virtual
+        returns (address[] memory)
+    {
+        return _cashbackRecipients[tokenId];
+    }
+
+    function updateCashbackForAuthor(uint256 tokenId, uint256 cashbackValue)
+        public
+        returns (bool)
+    {
+        for (uint256 i = 0; i < _cashbackValues[tokenId].length; i++) {
+            if (_cashbackRecipients[tokenId][i] == _msgSender()) {
+                _cashbackValues[tokenId][i] = cashbackValue;
+                return true;
+            }
+        }
+        return true;
+    }
+
+    function burn(uint256 tokenId) public virtual {
+        //solhint-disable-next-line max-line-length
+        require(
+            _isApprovedOrOwner(_msgSender(), tokenId),
+            "ERC721Burnable: caller is not owner nor approved"
+        );
+        _burn(tokenId);
+    }
+
+    function safeTransfer(
+        address to,
+        uint256 tokenId,
+        string memory data,
+        uint256 value
```

Comment:
cant we remove this value from the parameter and move it to the data? so we will have data and value in 1 field separated by some our internal ####""""""##### or something like that? and we will split it here and we will be compatible with the safeTransfer operation